Objects/stringobject.c

```diff
@@ -1029,6 +1029,15 @@ string_concat(register PyStringObject *a, register PyObject *bb)
 {
     register Py_ssize_t size;
     register PyStringObject *op;
+    if (a->ob_type != bb->ob_type) {
+        if (Py_Py3kWarningFlag &&
+            PyErr_WarnEx(PyExc_Py3xWarning,
+                    "string concatenation on mixed types is not supported in 3.x:"
```

Comment:
How about something like:

> mixed `bytes`, `string` and `unicode` operands cannot be used in string concatenation in Python 3.x: convert the operand(s) so that they are the same type.